src/components/utils/src/pipe_qt.cc

```diff
@@ -0,0 +1,174 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <string>
+#include <algorithm>
+
+#include "utils/winhdr.h"
+#include <QtNetwork>
+#include "utils/pipe.h"
+
+namespace std {
+
+template <>
+void swap<utils::Pipe>(utils::Pipe& lhs, utils::Pipe& rhs) {
+  lhs.Swap(rhs);
+}
+
+}  // namespace std
+
+namespace utils {
+
+class Pipe::Impl {
+ public:
+  Impl();
+  explicit Impl(QLocalSocket pipe);
+  ~Impl();
+
+  Impl(Impl& rh);
+  Impl& operator=(Impl& rh);
+
+  bool Valid() const;
+
+  bool Create(const std::string& name);
+
+  bool Open();
+  bool Close();
+
+  ssize_t Write(const char* buf, size_t length);
+
+ private:
+  QLocalSocket* pipe_;
+  QString name_;
+};
+
+}  // namespace utils
+
+utils::Pipe::Pipe() : impl_(new Pipe::Impl()) {}
+
+utils::Pipe::~Pipe() { delete impl_; }
+
+utils::Pipe::Pipe(Pipe& rh) {
+  impl_ = new Pipe::Impl();
+  *impl_ = *rh.impl_;
+}
+
+utils::Pipe& utils::Pipe::operator=(Pipe& rh) {
+  if (this != &rh) {
+    Pipe tmp(rh);
+    this->Swap(tmp);
+  }
+  return *this;
+}
+
+bool utils::Pipe::Valid() const { return impl_->Valid(); }
+
+bool utils::Pipe::Create(const std::string& name) {
+  return impl_->Create(name);
+}
+
+bool utils::Pipe::Open() { return impl_->Open(); }
+
+bool utils::Pipe::Close() { return impl_->Close(); }
+
+ssize_t utils::Pipe::Write(const char* buf, size_t length) {
+  return impl_->Write(buf, length);
+}
+
+utils::Pipe::Pipe(Pipe::Impl* impl) : impl_(impl) {}
+
+void utils::Pipe::Swap(Pipe& rh) { std::swap(this->impl_, rh.impl_); }
+
+utils::Pipe::Impl::Impl() : pipe_(NULL) {}
+
+utils::Pipe::Impl::Impl(QLocalSocket pipe) : pipe_(NULL) {}
+
+utils::Pipe::Impl::~Impl() { Close(); }
+
+utils::Pipe::Impl::Impl(Impl& rh) : pipe_(rh.pipe_) {
+  rh.pipe_ = NULL;
+  rh.name_ = "";
+}
+
+utils::Pipe::Impl& utils::Pipe::Impl::operator=(Impl& rh) {
+  Close();
+  pipe_ = rh.pipe_;
+  rh.pipe_ = NULL;
+  rh.name_ = "";
+  return *this;
+}
+
+bool utils::Pipe::Impl::Valid() const { return pipe_ != NULL; }
+
+bool utils::Pipe::Impl::Create(const std::string& name) {
+  if (!Close()) {
+    return false;
+  }
+  pipe_ = new QLocalSocket(QCoreApplication::instance());
+  if (!pipe_) {
+    pipe_ = NULL;
+    return false;
+  }
+  QString stream_path(name.c_str());
+  QStringList stream_path_list = stream_path.split(QDir::separator());
+  name_ = "\\\\.\\pipe\\";
+  name_.append(stream_path_list.at(stream_path_list.size() - 1));
+  return true;
+}
+
+bool utils::Pipe::Impl::Open() {
+  pipe_->setReadBufferSize(NULL);
+  return true;
+}
+
+bool utils::Pipe::Impl::Close() {
+  if (NULL == pipe_) {
+    return true;
+  }
+  pipe_->abort();
+  delete pipe_;
+  pipe_ = NULL;
+  name_ = "";
+  return true;
+}
+
+ssize_t utils::Pipe::Impl::Write(const char* buf, size_t length) {
+  if (NULL == pipe_) {
+    return -1;
+  }
+  pipe_->abort();
+  pipe_->connectToServer(name_, QIODevice::WriteOnly);
```

Comment:
`Pipe::Open` method is designed for connecting to pipe. Maybe move this logic there and do not reconnect on each write attempt?